### ISSUE TITLE
noqa is ignored in custom rule on fix

### ISSUE THREAD

**Issue by sti0** [sqlfluff/sqlfluff#1454]
_If this is a parsing or linting issue, please include a minimal SQL example which reproduces the issue, along with the `sqlfluff parse` output, `sqlfluff lint` output and `sqlfluff fix` output when relevant._

## Expected Behaviour
Respect `noqa` in the mentioned line when fixing the SQL file.

## Observed Behaviour
`noqa` is not respected when I run `sqlfluff fix --dialect=exasol`.

## Steps to Reproduce
I created this custom rule, which removes double quotes from a CodeSegment

```
@document_fix_compatible
class Rule_EXAMPLE_R002(BaseRule):
    """Don't use double quotes. """

    def __init__(self, *args, **kwargs):
        """Overwrite __init__ to set config."""
        super().__init__(*args, **kwargs)

    def _eval(self, segment, raw_stack, **kwargs):
        fixes = []
        if segment.name == "quoted_identifier" and isinstance(segment, CodeSegment):
            fixes.append(
                LintFix("edit", segment, segment.edit(segment.raw.replace('"', "")))
            )

        if fixes:
            return LintResult(
                anchor=segment,
                description="Don't use double quotes.",
                fixes=fixes,
            )

```

This is my testfile:
```
CREATE TABLE "SANDBOX"."TAB1" (
      "col1" VARCHAR(200)
      "KEY" VARCHAR(200) -- noqa
    );
```
If I `lint` it:
```
L:   1 | P:  14 | EXAMPLE_R002 | Don't use double quotes.
L:   1 | P:  24 | EXAMPLE_R002 | Don't use double quotes.
L:   2 | P:   7 | L003 | Indentation not hanging or a multiple of 4 spaces
L:   2 | P:   7 | EXAMPLE_R002 | Don't use double quotes.
L:   4 | P:   5 | L003 | Indentation not consistent with line #1
L:   4 | P:   6 | L009 | Files must end with a trailing newline.
```
Please note that line 3 (`KEY`, noqa line) is not printed (as expected).

After running the `fix` command:
```
CREATE TABLE IF NOT EXISTS SANDBOX.TAB1 (
    col1 VARCHAR(200)
    KEY VARCHAR(200) -- noqa
);
```
==> double quotes on KEY are removed.

## Dialect
exasol

## Version
Include the output of `sqlfluff --version` along with your Python version
sqlfluff, version 0.6.6
Python 3.8.12

## Configuration
default

**Comment by tunetheweb** [comment]
I think this is the same as #1471 

Can you confirm it works with: `sqlfluff fix --dialect=exasol --exclude-rules EXAMPLE_R002 `
And also when excluding via `.sqlfluff`

**Comment by sti0** [comment]
Sure, if I exclude this rule the quotes resist. But I like to use this rule to remove the double quotes (except the one in the qa line).

BUT if I add a `--noqa: disable=EXAMPLE_R002` to the **top** of the file:
```
--noqa: disable=EXAMPLE_R002
CREATE TABLE "SANDBOX"."TAB1" (
      "col1" VARCHAR(200),
      "KEY" VARCHAR(200)
    );
```

 the lint output shows the following:
```
=== finding fixable violations ====
WARNING    One fix for EXAMPLE_R002 not applied, it would re-cause the same error. 
== [quotetest.sql] FAIL
L:   3 | P:   7 | L003 | Indentation not hanging or a multiple of 4 spaces
L:   4 | P:   7 | L003 | Indentation not hanging or a multiple of 4 spaces
L:   5 | P:   5 | L003 | Indentation not consistent with line #2
L:   5 | P:   6 | L009 | Files must end with a trailing newline.
```
==> as you could see, the rule is not applied. But results in:
```
--noqa: disable=EXAMPLE_R002
CREATE TABLE IF NOT EXISTS SANDBOX.TAB1 (
    col1 VARCHAR(200),
    KEY VARCHAR(200)
);
```
But the same happens when I ignore L003. This is strange....this happens even if I setup sqlfluff in a clean venv.


Even this doesn't work:
```
CREATE TABLE "SANDBOX"."TAB1" (
      "col1" VARCHAR(200),
      --noqa: disable=EXAMPLE_R002
      "KEY" VARCHAR(200),
      --noqa: enable=all
      "id" NUMBER
    );
```

I think there is something broken with noqa & fix.

**Comment by tunetheweb** [comment]
> the lint output shows the following:

`lint` doesn't seem to suffer from this bug and `fix` basically runs lint first and that's the output you see. If you run `fix` with `-vvvv` you can see that the rule is applied. Which I agree is wrong.

> I think there is something broken with noqa & fix.

Yup!

**Comment by barrywhart** [comment]
Noqa is designed to run all the rules and then filter out any violations of the designated rules. In other words, the fact it runs a "noqa" rule is fine.

**Comment by sti0** [comment]
I think this ist very misleading. As the tickets of the last days showed there are some people who expected that fix respects noqa. If Not some usecases (e.g. pre-commit) are limited. We have to exclude the entire rule and can not handle special situations by exclude a rule by file or line.

I think it's worth to overthink that behaviour.

**Comment by tunetheweb** [comment]
I don’t think @barrywhart is saying that what you’re experiencing is intended or accepted.

I think he was addressing my comment to let me know that seeing it running is not necessarily wrong. But yes it shouldn’t then act upon that run.

So think we all agree `fix` should honour `noqa` directives and doesn’t currently honour inline `noqa` directives when it should.

**Comment by barrywhart** [comment]
We have some tests for the inline filtering. I wonder if there are other factors. For example, if the parser gets the line numbers wrong, that can make the filtering misbehave. That mainly happens with templated code. Just mentioning it as something to be aware of when we look into this.

**Comment by tunetheweb** [comment]
@sti0 could you try this again on `main` now that #1502 is merged? Or in 0.6.7 which is soon to be released!

I\m goign to close this assuming that fixed that too, but feel free to reopen if that's not the case!

**Comment by sti0** [comment]
I can confirm that this issue is solved by #1502 . Thanks @barrywhart
